### ISSUE TITLE
BVH is causing MeshInstance light to not update automatically in GLES2 when visibility is changed

### ISSUE THREAD

**Issue by filipworkson** [godotengine/godot#45292]
**Godot version:**

3.2.4 beta 6

**OS/device including version:**

Windows 10
GLES2
Nvidia GTX 1650 Super

**Issue description:**

I use a custom light shader on my billboarded items to create an even light look. 

```
void light() {
    DIFFUSE_LIGHT += clamp(ATTENUATION * ALBEDO,0,0.5);
}
```

BVH
The light doesn't update if I change an object visibility until light in the scene changes a bit. My player has a light attached to it and by moving closer or further away from the item I trigger a lighting update which makes my light shader work again.  
![bvh](https://user-images.githubusercontent.com/36826319/104955102-9fbf9a00-5997-11eb-88d0-6df7f704961e.gif)

Without BVH
Here is without bvh. As you can see the light updates automatically  after changing object visibility.
![without_bvh](https://user-images.githubusercontent.com/36826319/104955189-cb428480-5997-11eb-85e2-a61f5b74207f.gif)

The item looks like this
![image](https://user-images.githubusercontent.com/36826319/104955868-e9f54b00-5998-11eb-9b22-a02e62456fa8.png)

When I pick up the item I simply hide it an change it's collision flag and disable physics then when I drop it I raycast to mose pos and place it at the collision point then enable physics and collision flag. I believe when a mesh visibility is changed BVH doesn't update until light setup changes in the scene. This causes the object to appear like it has no light on it.

**Steps to reproduce:**
This is always reproduceable in my game. Someone needs to confirm this.

**Minimal reproduction project:**

**Comment by Calinou** [comment]
This is likely fixed by https://github.com/godotengine/godot/commit/694e51d210b1a376418c9d067aa9d763689b5770, please test by using the latest GitHub Actions builds: https://github.com/godotengine/godot/actions?query=branch%3A3.2+event%3Apush

**Comment by lawnjelly** [comment]
Yes this does sound similar to #45246.

**Comment by filipworksworks** [comment]
I have tested the new version and this issue is mostly gone. It does show up again on occasion but is much more rare. It seems to show up more often right after running the project or resizing the window.

**Comment by lawnjelly** [comment]
> I have tested the new version and this issue is mostly gone. It does show up again on occasion but is much more rare. It seems to show up more often right after running the project or resizing the window.

Good for the improvement.

This could still be possible (especially in the editor). The old system I'm presuming would update collisions as each object is moved, and there is no update function, whereas the new system has a separate collision detection step run on all the objects at once, which is run when the BVH is updated. This is more efficient.

The initial assumption was that the update function from the render tree would be called fairly regularly or whenever something changed ... however in the editor it turns out this isn't the case for the render tree in all situations. So part of the previous fix PR was forcing the collision detection step manually whenever the pairing is changed in the editor, which happens when you change the visibility of an item.

It may be there still some more occasions where update should be called and isn't. I'll try and have a look and work out. The good news is this is nothing sinister, just a lack of an update, and these are now seeming rare and it should not affect games when run. Do let me know any minimum reproduction projects or circumstances you notice it on, that will be easier to fix.

**Comment by filipworksworks** [comment]
@lawnjelly Thank you for the reply and for your work!  I only noticed one edge case was right at the start of the game probably when is still building the bvh?  I tested it consistently and doesn't happen past that point. In fact it fixes itself soon after game start and the issue doesn't happen again. 

I will try again with your pr when is gonna get merged.  

Should this issue be closed? I will let you guys decide. I for one am very happy with the results after the changes!

**Comment by lawnjelly** [comment]
I think we can keep the issue open as there are still edge cases. It is almost inevitable, however we should be able to fix all of them I think. :+1: 

You can actually try the PR before merging - we have had this ability for a few weeks now, you can use it for any PR.

* just follow this link:
https://github.com/godotengine/godot/pull/45311/commits
* Click on the green arrow
* Click on details for windows or linux build
* Click on artifacts in the top right
* Here you will see links to download the editor

**Comment by fractilegames** [comment]
I assume the the fix applies to GLES3 as well? I'm seeing very similar behavior in beta6 using GLES3 renderer: https://gfycat.com/illegalbrownaxisdeer

**Comment by lawnjelly** [comment]
> I assume the the fix applies to GLES3 as well? I'm seeing very similar behavior in beta6 using GLES3 renderer: https://gfycat.com/illegalbrownaxisdeer

Yup, spatial partitioning happens before any of the rasterizers.

**Comment by filipworksworks** [comment]
I have tested the new version with the new pr and is actually slightly worse than  694e51d. 

Before it used to happen only once right after I start the game then didn't happen at all after that.

With the new build it happens a number of times and I have to move the object around a few times before it goes away and light works normally. Is deff something to do with the way my code works.

Here is the code for how I grab and release an item. 
![image](https://user-images.githubusercontent.com/36826319/105531709-bf1c3700-5cb7-11eb-8554-b400fc9413e4.png)

When I grab 
I change the collision layer so it no longer collides with anything. I had to do this because of another bug where disabling physics doesn't work.
I remove the item from the `things` node which is a generic Node I use as a container
I place the item into player's `items` node which is another generic Node node I use as a list of items

When  I release I do the same but in reverse

Note that player.grabbed_item.global_transform.origin I set the position BEFORE I do all that above. I think this might be the issue?

**Comment by lawnjelly** [comment]
There was another PR since that one, it turned out there was a problem with one of the functions it relied on.
But essentially, in order to fix bugs, the long and short is I need a minimum reproduction project, otherwise it is just guesswork.

**Comment by filipworksworks** [comment]
My current project is very big and has a lot of scripts and files. I would need to strip off only those elements so you can test them out. It's a little bit of work. I will see what I can do.

**Comment by akien-mga** [comment]
This might be fixed by #45530 if you can test. You can try a CI build from the `3.2` branch to see: https://github.com/godotengine/godot/runs/1792278268

(Currently building but once done you should be able to download artifacts.)

**Comment by filipworkson** [comment]
For me personally this issue is now fixed. I have not seen this in latest 3.2.3 rc3 and is now working with bvh same way as before.  

Thank you very much!
